File: 79.word-search.java
public class Solution {
	public boolean exist(char[][] board, String word) {
		for(int i=0;i<board.length;i++){
			for(int j=0;j<board[i].length;j++){
				if(helper(board,i,j,word,0)){
					return true;
				}
			}
		}
		return false;
	}
	private boolean helper(char[][] board, int i, int j, String word, int step){
		if(step==word.length()){
			return true;
		}
		if(i<0||i>=board.length||j<0||j>=board[i].length){
			return false;
		}
		//开头挂就是直接挂
		if((board[i][j]-word.charAt(step))!=0){
			return false;
		}
		
		char record=board[i][j];
        board[i][j]='1'; //防止 i-1 或 j-1等等这些回来时发现又相等 比如【“aa”】,aaa 这个到第三个a，又回到【“aa”】中的第一个a了，防止回调
		boolean res=helper(board,i-1,j,word,step+1)
				||helper(board,i,j-1,word,step+1)
				||helper(board,i,j+1,word,step+1)
				||helper(board,i+1,j,word,step+1);
		board[i][j]=record;
		return res;
	}
}

File: 79.word-search.js
/*
 * @lc app=leetcode id=79 lang=javascript
 *
 * [79] Word Search
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
  if (board.length === 0) return false;
  const [row, column] = [board.length, board[0].length];
  const dirs = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
  ];

  // 坐标 + 看第哪一个word
  const dfs = (x, y, k) => {
    if (board[x][y] !== word[k]) return false;
    if (k === word.length - 1) return true;

    board[x][y] = "*"; // mark as visited
    for (const [dx, dy] of dirs) {
      const i = x + dx;
      const j = y + dy;
      if (i >= 0 && i < row && j >= 0 && j < column) {
        if (dfs(i, j, k + 1)) return true;
      }
    }

    board[x][y] = word[k]; // reset
    return false;
  };

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < column; j++) {
      if (board[i][j] === word[0] && dfs(i, j, 0)) return true;
    }
  }

  return false;
};
// @lc code=end
